x/evm/keeper/vote_handler.go

```diff
@@ -12,65 +12,153 @@ import (
 	"github.com/axelarnetwork/utils/slices"
 )
 
+var _ vote.VoteHandler = &voteHandler{}
+
+type voteHandler struct {
+	cdc      codec.Codec
+	keeper   types.BaseKeeper
+	nexus    types.Nexus
+	rewarder types.Rewarder
+}
+
 // NewVoteHandler returns the handler for processing vote delivered by the vote module
-func NewVoteHandler(cdc codec.Codec, keeper types.BaseKeeper, nexus types.Nexus) vote.VoteHandler {
-	return func(ctx sdk.Context, poll vote.Poll) error {
-		voteEvents, err := types.UnpackEvents(cdc, poll.GetResult().(*vote.Vote).Result)
-		if err != nil {
-			return err
-		}
+func NewVoteHandler(cdc codec.Codec, keeper types.BaseKeeper, nexus types.Nexus, rewarder types.Rewarder) vote.VoteHandler {
+	return voteHandler{
+		cdc:      cdc,
+		keeper:   keeper,
+		nexus:    nexus,
+		rewarder: rewarder,
+	}
+}
 
-		if slices.Any(voteEvents.Events, func(event types.Event) bool { return event.Chain != voteEvents.Chain }) {
-			return fmt.Errorf("events are not from the same source chain")
-		}
+func (v voteHandler) IsFalsyResult(result codec.ProtoMarshaler) bool {
+	voteEvents, err := types.UnpackEvents(v.cdc, result.(*vote.Vote).Result)
 
-		chain, ok := nexus.GetChain(ctx, voteEvents.Chain)
-		if !ok {
-			return fmt.Errorf("%s is not a registered chain", voteEvents.Chain)
-		}
+	return err != nil || len(voteEvents.Events) == 0
+}
+
+func (v voteHandler) HandleExpiredPoll(ctx sdk.Context, poll vote.Poll) error {
+	rewardPoolName, ok := poll.GetRewardPoolName()
+	if !ok {
+		return fmt.Errorf("reward pool not set for poll %s", poll.GetKey().String())
+	}
 
-		if !keeper.HasChain(ctx, voteEvents.Chain) {
-			return fmt.Errorf("%s is not an evm chain", voteEvents.Chain)
+	// TODO: MarkChainMaintainerMissingVote for those who didn't vote in time. Need
```

Comment:
One hack would be to use the reward pool name which corresponds to the chain name, but since this scenario is uncommon, it's better to wait until a proper fix.

Reply:
Let's wait for it. I thought of the hack too and decided not to rush it since it is confusing.